File: test-local-pkg/ts/tests/test-frontend-sdk.ts
import { pathToFileURL } from 'url';
import { client, runTestSection, logResult, organizationId } from "../utils/utils.js";
import { randomUUID } from "crypto";

/**
 * TEST FRONTEND SDK (SCENARIO-BASED)
 * Focuses on Chat Client API, Client Messages, Votes, and Tracing.
 */
async function testFrontendSdk() {
  console.log("\n" + "=".repeat(70));
  console.log("🚀 STARTING: FRONTEND SDK & TRACING TEST");
  console.log("=".repeat(70));

  const ts = Date.now();
  const state = {
    chatId: null as string | null,
    userId: null as string | null,
    messageId: null as string | null,
    orgId: organizationId as string || "",
  };

  try {
    // --- Section 1: System Info & Health ---
    await runTestSection("System Info & Health", async () => {
      const config = await client.aiAgent.getConfig();
      logResult("Config", config);

      const health = await client.aiAgent.getHealth(true);
      logResult("Health", health);

      const version = await client.aiAgent.getVersion();
      logResult("Version", version);
    }).catch(e => console.error(`\n❌ Section [System Info] Failed: ${e.message}`));

    // --- Section 2: Chat Client Auth ---
    await runTestSection("Chat Client Auth", async () => {
      // Register (or get) a chat client user
      const email = `sdk-test-${ts}@example.com`;
      const password = "Password123!";
      
      const user = await client.aiAgent.registerChatClient({ 
          id: `sdk-test-user-${ts}`,
          email,
          password
      }).catch(async () => {
          return await client.aiAgent.getChatClientUser({ id: `sdk-test-user-${ts}` });
      });
      
      state.userId = user.id;
      logResult("Chat Client User", state.userId);

      const verified = await client.aiAgent.verifyChatClientCredentials({ 
          email,
          password
      });
      logResult("Credentials Verified", !!verified);
    }).catch(e => console.error(`\n❌ Section [Chat Client Auth] Failed: ${e.message}`));

    // --- Section 3: Chat Client Chats ---
    await runTestSection("Chat Client Chats", async () => {
      if (!state.userId) {
          console.log("   ⚠️ Skipping Chat Client Chats because userId is null.");
          return;
      }
      
      const assistants = await client.chatAi.listAssistants();
      const realAsstId = assistants.length > 0 ? assistants[0].id : "dummy-id";

      const chat = await client.aiAgent.createClientChat({
        id: randomUUID(),
        assistantId: realAsstId,
        organizationId: state.orgId,
        userId: state.userId!,
        selectedVisibilityType: "public",
        message: {
          id: randomUUID(),
          role: "user",
          content: "Hello from Frontend SDK Test",
          createdAt: new Date().toISOString(),
          parts: [{ type: "text", text: "Hello from Frontend SDK Test" }]
        }
      });
      state.chatId = chat._id || chat.id;
      logResult("Client Chat Created", state.chatId);

      const fetchedChat = await client.aiAgent.getClientChat(state.chatId!);
      logResult("Get Client Chat", fetchedChat.id);

      const chats = await client.aiAgent.listClientChats({ limit: 5, organization_id: state.orgId });
      logResult("List Client Chats", chats.data?.length);

      try {
          const titleRes = await client.aiAgent.generateClientChatTitle(state.chatId!);
          logResult("Generated Title", titleRes.title);
      } catch {}

      await client.aiAgent.updateClientChat(state.chatId!, { visibility: "private" });
      logResult("Updated Visibility", "private");

      // 4. Create Document (Frontend)
      const doc = await client.aiAgent.createDocument({
          kind: "text",
          content: "Sample document content created from SDK test",
          title: "SDK Test Doc"
      });
      logResult("Frontend Document Created", doc._id || doc.id);
    }).catch(e => console.error(`\n❌ Section [Chat Client Chats] Failed: ${e.message}`));

    // --- Section 4: Client Messages & Votes ---
    await runTestSection("Client Messages & Votes", async () => {
      if (!state.chatId) {
          console.log("   ⚠️ Skipping Messages & Votes because chatId is null.");
          return;
      }
      
      const msg = await client.aiAgent.persistClientMessage({
        chatId: state.chatId!,
        role: "assistant",
        content: "Hello! How can I help you today?",
        parts: [{ type: "text", text: "Hello! How can I help you today?" }]
      });
      state.messageId = msg._id || msg.id;
      logResult("Message Persisted", state.messageId);

      const messages = await client.aiAgent.listClientMessages(state.chatId!);
      logResult("List Messages", messages.length);

      await client.aiAgent.updateVote({
        chatId: state.chatId!,
        messageId: state.messageId!,
        vote: 1 // Upvote
      });
      logResult("Vote Updated", "Upvote");

      const votes = await client.aiAgent.getVotes(state.chatId!);
      logResult("Votes Count", votes.length);
    }).catch(e => console.error(`\n❌ Section [Client Messages & Votes] Failed: ${e.message}`));

    // --- Section 5: Tracing (Tempo) ---
    await runTestSection("Tracing (Tempo)", async () => {
      const services = await client.aiAgent.getTraceServices();
      logResult("Trace Services", services);

      const tags = await client.aiAgent.getTraceTags();
      logResult("Trace Tags", tags);

      const traces = await client.aiAgent.getTraces({ limit: 5, orgId: state.orgId });
      logResult("Traces Found", traces?.traces?.length);

      if (traces?.traces?.length > 0) {
        const traceId = traces.traces[0].traceID;
        const details = await client.aiAgent.getTrace(traceId);
        logResult("Trace Details", traceId);
      }
    }).catch(e => console.error(`\n❌ Section [Tracing] Failed: ${e.message}`));

    console.log("\n" + "=".repeat(70));
    console.log("✅ FRONTEND SDK TEST COMPLETED SUCCESSFULLY");
    console.log("=".repeat(70));

  } catch (err: any) {
    console.error("\n❌ FATAL TEST ERROR:", err.message);
    throw err;
  } finally {
    console.log("\n[Cleanup] Cleaning up chat client resources...");
    try {
      if (state.chatId) await client.aiAgent.deleteClientChat(state.chatId);
      console.log("   Cleanup finished.");
    } catch (e: any) {
      console.warn("   Cleanup failed:", e.message);
    }
  }
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  testFrontendSdk().catch(console.error);
}

export { testFrontendSdk };

File: test-npm-pip-pkg/py/tests/test_frontend_sdk.py
import time
import uuid
from utils.utils import client, run_test_section, log_result, organization_id

def test_frontend_sdk():
    print("\n🚀 STARTING: FRONTEND SDK & TRACING TEST (PY)")

    state = {"chat_id": None, "user_id": None}

    # 1. Chat Client Auth
    def chat_client_auth():
        ts = int(time.time())
        email = f"sdk-py-test-{ts}@example.com"
        password = "Password123!"

        try:
            user = client.ai_agent.register_chat_client({
                "id": f"sdk-py-user-{ts}",
                "email": email,
                "password": password
            })
        except:
            user = client.ai_agent.get_chat_client_user({"id": f"sdk-py-user-{ts}"})
        
        state["user_id"] = user.get("id")
        log_result("Chat Client User", state["user_id"])

        verified = client.ai_agent.verify_chat_client_credentials({
            "email": email,
            "password": password
        })
        log_result("Credentials Verified", bool(verified))
    run_test_section("Chat Client Auth", chat_client_auth)

    # 2. Chat Client Chats
    def chat_client_chats():
        if not state["user_id"]: return
        
        assistants = client.chat_ai.list_assistants()
        asst_id = assistants[0]["id"] if assistants else "dummy-id"

        chat = client.ai_agent.create_client_chat({
            "id": str(uuid.uuid4()),
            "assistant_id": asst_id,
            "organization_id": organization_id,
            "user_id": state["user_id"],
            "selectedVisibilityType": "public",
            "message": {
                "id": str(uuid.uuid4()),
                "role": "user",
                "content": "Hello from Python Frontend SDK",
                "createdAt": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
                "parts": [{"type": "text", "text": "Hello"}]
            }
        })
        state["chat_id"] = chat.get("_id") or chat.get("id")
        log_result("Client Chat Created", state["chat_id"])

        if state["chat_id"]:
            fetched = client.ai_agent.get_client_chat(state["chat_id"])
            log_result("Get Client Chat", fetched.get("id"))

            client.ai_agent.update_client_chat(state["chat_id"], {"visibility": "private"})
            log_result("Updated Visibility", "private")
            
            # --- Chat Client — Messages ---
            try:
                client.ai_agent.persist_client_message({"chat_id": state["chat_id"], "message": {"role": "user", "content": "hello"}})
            except Exception as e:
                pass
            try:
                client.ai_agent.list_client_messages(state["chat_id"])
            except Exception as e:
                pass
            try:
                client.ai_agent.delete_trailing_messages("dummy_msg_id")
            except Exception as e:
                pass

            # --- Chat Client — Votes ---
            try:
                client.ai_agent.get_votes(state["chat_id"])
            except Exception as e:
                pass
            try:
                client.ai_agent.update_vote({"chat_id": state["chat_id"], "message_id": "dummy_msg_id", "vote": 1})
            except Exception as e:
                pass

            # --- Chat Client — Misc ---
            try:
                client.ai_agent.stream_client_chat_status(state["chat_id"])
            except Exception as e:
                pass
            try:
                client.ai_agent.generate_client_chat_title(state["chat_id"])
            except Exception as e:
                pass

            # Cleanup
            client.ai_agent.delete_client_chat(state["chat_id"])
            print("   Chat deleted.")

        # Cleanup all
        try:
            client.ai_agent.delete_all_client_chats(organization_id)
        except Exception as e:
            pass
    run_test_section("Chat Client Chats", chat_client_chats)

    # 3. Chat Client Documents
    def chat_client_documents():
        try:
            client.ai_agent.create_document({"chat_id": "dummy", "content": "test"})
        except Exception as e:
            pass
        try:
            client.ai_agent.get_document_latest_by_kind(chat_id="dummy", kind="summary")
        except Exception as e:
            pass
        try:
            client.ai_agent.get_document("dummy_doc_id")
        except Exception as e:
            pass
        try:
            client.ai_agent.get_document_latest("dummy_doc_id")
        except Exception as e:
            pass
        try:
            client.ai_agent.get_document_public("dummy_doc_id")
        except Exception as e:
            pass
        try:
            client.ai_agent.get_document_suggestions("dummy_doc_id")
        except Exception as e:
            pass
        try:
            client.ai_agent.delete_document("dummy_doc_id")
        except Exception as e:
            pass
    run_test_section("Chat Client Documents", chat_client_documents)

    print("\n✅ Frontend SDK Test Completed.")

if __name__ == "__main__":
    test_frontend_sdk()
